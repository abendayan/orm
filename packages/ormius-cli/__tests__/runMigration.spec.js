let mockError = null
let mockErrorSecond = null

const mockCallback = (query, callback) => {
    if (query === 'CREATE TABLE IF NOT EXISTS migrations (migration VARCHAR(255))') {
        callback(mockError, [])
    } else if (query === 'SELECT migration FROM migrations') {
        callback(mockErrorSecond, [{ migration: 'test' }])
    }
}

const mockOrmius = {
    connection: {
        query: jest.fn((query, callback) => mockCallback(query, callback))
    }
}

const { runMigration, executeMigration } = require('../cli/runMigration')

const mockMigration = {
    setConnection: jest.fn(),
    setMigrationId: jest.fn(),
    changes: jest.fn(),
    markAsRan: jest.fn()
}

jest.mock('fs', () => ({
    readdirSync: jest.fn(() => {
        const directory = { name: 'migration' }

        directory.isDirectory = () => true
        return [directory]
    })
}))

jest.mock('../cli/migrations/test/index', () => (
    jest.fn(() => mockMigration)
), { virtual: true })

jest.mock('../cli/migrations/migration/index', () => (
    jest.fn(() => mockMigration)
), { virtual: true })

jest.mock('ormius/lib/ormius', () => ({
    Orm: jest.fn(() => mockOrmius)
}))

describe('run migration test', () => {
    test('expect argument', () => {
        try {
            runMigration()
            expect(false).toBe(true)
        } catch (e) {
            expect(e.message).toBe('Missing argument config')
        }
    })

    test('executeMigration', () => {
        const mockConnection = {}

        executeMigration(mockConnection, 'test')
        expect(mockMigration.setConnection).toHaveBeenCalledWith(mockConnection)
        expect(mockMigration.setMigrationId).toHaveBeenCalledWith('test')
        expect(mockMigration.changes).toHaveBeenCalled()
        expect(mockMigration.markAsRan).toHaveBeenCalled()
    })

    test('runMigration', () => {
        runMigration('config.json')
        expect(mockOrmius.connection.query.mock.calls).toMatchSnapshot()
    })

    test('runMigration with error', () => {
        mockError = 'Error'
        try {
            runMigration('config.json')
            expect(false).toBe(true)
        } catch (e) {
            expect(e).toBe('Error')
        }
    })

    test('runMigration with error', () => {
        mockError = null
        mockErrorSecond = 'Error'
        try {
            runMigration('config.json')
            expect(false).toBe(true)
        } catch (e) {
            expect(e).toBe('Error')
            expect(mockOrmius.connection.query.mock.calls).toMatchSnapshot()
        }
    })
})
