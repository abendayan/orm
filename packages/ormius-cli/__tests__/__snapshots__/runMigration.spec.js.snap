// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`run migration test runMigration 1`] = `
Array [
  Array [
    "CREATE TABLE IF NOT EXISTS migrations (migration VARCHAR(255))",
    [Function],
  ],
  Array [
    "SELECT migration FROM migrations",
    [Function],
  ],
]
`;

exports[`run migration test runMigration with error 1`] = `
Array [
  Array [
    "CREATE TABLE IF NOT EXISTS migrations (migration VARCHAR(255))",
    [Function],
  ],
  Array [
    "SELECT migration FROM migrations",
    [Function],
  ],
  Array [
    "CREATE TABLE IF NOT EXISTS migrations (migration VARCHAR(255))",
    [Function],
  ],
  Array [
    "CREATE TABLE IF NOT EXISTS migrations (migration VARCHAR(255))",
    [Function],
  ],
  Array [
    "SELECT migration FROM migrations",
    [Function],
  ],
]
`;
